src/raftstore/store/snap/reader.rs

```diff
@@ -0,0 +1,274 @@
+use std::fs::File;
+use std::io::{self, BufReader, Read};
+use std::mem;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use rocksdb::{IngestExternalFileOptions, Writable, WriteBatch};
+
+use raftstore::store::keys;
+use raftstore::store::metrics::*;
+use raftstore::store::snap::*;
+use raftstore::store::util::check_key_in_region;
+use util::codec::bytes::CompactBytesFromFileDecoder;
+use util::file::{calc_crc32, delete_file_if_exist, get_file_size};
+use util::rocksdb::{get_cf_handle, prepare_sst_for_ingestion};
+
+struct Inner {
+    dir: String,
+    for_send: bool,
+    key: SnapKey,
+    snapshot_meta: SnapshotMeta,
+}
+
+impl Inner {
+    fn new(dir: String, for_send: bool, key: SnapKey, snapshot_meta: SnapshotMeta) -> Self {
+        Inner {
+            dir,
+            for_send,
+            key,
+            snapshot_meta,
+        }
+    }
+}
+
+pub struct SnapshotSender {
+    inner: Inner,
+    snap_stale_notifier: Arc<SnapStaleNotifier>,
+    ref_count: Arc<AtomicUsize>,
+    sent_times: Arc<AtomicUsize>,
+
+    cf_file: Option<File>,
+    cf_index: usize,
+}
+
+impl SnapshotSender {
+    pub(super) fn new(
+        dir: String,
+        key: SnapKey,
+        snapshot_meta: SnapshotMeta,
+        snap_stale_notifier: Arc<SnapStaleNotifier>,
+        ref_count: Arc<AtomicUsize>,
+        sent_times: Arc<AtomicUsize>,
+    ) -> Self {
+        let inner = Inner::new(dir, true, key, snapshot_meta);
+        ref_count.fetch_add(1, Ordering::SeqCst);
+        SnapshotSender {
+            inner,
+            snap_stale_notifier,
+            ref_count,
+            sent_times,
+            cf_file: None,
+            cf_index: 0,
+        }
+    }
+
+    pub fn total_size(&self) -> u64 {
+        get_size_from_snapshot_meta(&self.inner.snapshot_meta)
+    }
+}
+
+impl Drop for SnapshotSender {
+    fn drop(&mut self) {
+        self.ref_count.fetch_sub(1, Ordering::SeqCst);
+        self.sent_times.fetch_add(1, Ordering::SeqCst);
+    }
+}
+
+impl Read for SnapshotSender {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+        let inner = &mut self.inner;
+        while self.cf_index < inner.snapshot_meta.get_cf_files().len() {
```

Comment:
Like SnapshotGenerator, it's better extract a function `shift_to_next_cf_if_need`